titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/KubeApiServerIntegrator.java

```diff
@@ -351,24 +351,49 @@ private void nodeUpdated(V1Node node) {
      * </ul>
      */
     private boolean isNodeOwnedByFenzo(V1Node node) {
+        if (isFarzoneNode(node)) {
+            logger.debug("Not owned by fenzo (farzone node): {}", node.getMetadata().getName());
+            return false;
+        }
+        if (!hasFenzoSchedulerTaint(node)) {
+            logger.debug("Not owned by fenzo (non Fenzo scheduler taint): {}", node.getMetadata().getName());
+            return false;
+        }
+
         List<V1Taint> taints = node.getSpec().getTaints();
         if (CollectionsExt.isNullOrEmpty(taints)) {
-            return !isFarzoneNode(node);
+            logger.debug("Owned by fenzo (no taint set): {}", node.getMetadata().getName());
+            return true;
         }
 
+        Set<String> toleratedTaints = mesosConfiguration.getFenzoTaintTolerations();
+        for (V1Taint taint : taints) {
+            String taintKey = taint.getKey();
+            if (!taintKey.equals(KubeConstants.TAINT_SCHEDULER) && !toleratedTaints.contains(taintKey)) {
+                logger.debug("Not owned by fenzo (non tolerable taint found): nodeId={}, taintKey={}", node.getMetadata().getName(), taintKey);
+                return false;
+            }
+        }
+
+        logger.debug("Owned by fenzo (all taints tolerated): {}", node.getMetadata().getName());
+        return true;
+    }
+
+    /**
+     * Returns true if there is {@link KubeConstants#TAINT_SCHEDULER} taint with {@link KubeConstants#TAINT_SCHEDULER_VALUE_FENZO} value
+     * or this taint is missing (no explicit scheduler taint == Fenzo).
```

Comment:
the code below seems to be returning `false` when there is no explicit scheduler taint (i.e. `schedulerTaintValues.size() < 1`). What is the expected behavior?

Some unit tests here would be nice.

Reply:
Yes, this is correct. If there is no taint, we assume the node belongs to fenzo (see the comment).
We do not have unit tests for this class. I tested this extensively manually on a running system.

Reply:
Let me rephrase: the comment here is not consistent with the behavior of the code on line 396. If a Node has taints, but none of them are `KubeConstants.TAINT_SCHEDULER` the return of this method will be `false` and the node will be considered not owned by Fenzo, which is the opposite of what the javadoc says.

Reply:
If what I said is correct and I'm not missing anything, one more argument in favor of having a unit test for the behavior here (how taints are being used and expected).

Reply:
It is a bug. I am adding a unit test for this function.